etcd-defrag/etcddefrag_controller.go

```diff
@@ -0,0 +1,294 @@
+	}
+
+	successfulDefrags := 0
+	var errs []error
+	for _, status := range endpointStatus {
+		member, err := getMemberFromStatus(etcdMembers, status)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		// Check each member's status which includes the db size on disk "DbSize" and the db size in use "DbSizeInUse"
+		// compare the % difference and if that difference is over the max diff threshold and also above the minimum
+		// db size we defrag the members state file. In the case where this command only partially completed controller
+		// can clean that up on the next sync. Having the db sizes slightly different is not a problem in itself.
+		if r.isEndpointBackendFragmented(member, status) {
+			fragmentedPercentage := checkFragmentationPercentage(status.DbSize, status.DbSizeInUse)
+			r.log.Info("Member is over defrag threshold", "name", member.Name, "URL", member.ClientURLs[0], "fragmentation percentage", fragmentedPercentage, "DBSize on disk", status.DbSize, "DBSize in use", status.DbSizeInUse, "leader", status.Leader)
+			if _, err := r.etcdClient.Defragment(ctx, member); err != nil {
+				// Defrag can timeout if defragmentation takes longer than etcdcli.DefragDialTimeout.
+				r.log.Error(err, "DefragController Defragment Failed", "member", member.Name, "ID", member.ID)
+				errMsg := fmt.Sprintf("failed defrag on member: %s, memberID: %x: %v", member.Name, member.ID, err)
+				errs = append(errs, fmt.Errorf(errMsg))
+				continue
+			}
+
+			r.log.Info("DefragController Defragment Success", "member", member.Name, "ID", member.ID)
+			successfulDefrags++
+
+			// Give cluster time to recover before we move to the next member.
+			if err := wait.Poll(
+				pollWaitDuration,
+				pollTimeoutDuration,
+				func() (bool, error) {
+					// Ensure defragmentation attempts have clear observable signal.
+					r.log.Info("Sleeping to allow cluster to recover before defragging next member", "waiting", r.defragWaitDuration)
+					time.Sleep(r.defragWaitDuration)
+
+					memberHealth, err := r.etcdClient.MemberHealth(ctx)
+					if err != nil {
+						r.log.Error(err, "Failed checking member health")
+						return false, nil
+					}
+					if !etcdcli.IsClusterHealthy(memberHealth) {
+						r.log.Info("Cluster member is unhealthy", "member status", memberHealth.Status())
+						return false, nil
+					}
+					return true, nil
+				}); err != nil {
+				errs = append(errs, fmt.Errorf("timeout waiting for cluster to stabilize after defrag: %w", err))
+			}
+		} else {
+			// no fragmentation needed is also a success
+			successfulDefrags++
+		}
+	}
+
+	if successfulDefrags != len(endpointStatus) {
+		r.numDefragFailures++
+		r.log.Info("DefragController Defragment Partial Failure", "successfully defragged", successfulDefrags, "of members", len(endpointStatus), "tries remaining", maxDefragFailuresBeforeDegrade-r.numDefragFailures)
+
+		// return all errors here for the sync loop to retry immediately
+		return errors.Join(errs...)
```

Comment:
as we integrate with the etcd-operator we'd want to signal error in hcp conditions, so maybe add a todo here.

Reply:
OK will do.